tfhe/src/integer/u256.rs

```diff
@@ -344,10 +454,189 @@ impl From<u128> for U256 {
+
+impl Numeric for U256 {
+    const BITS: usize = Self::BITS as usize;
+
+    const ZERO: Self = Self::ZERO;
+
+    const ONE: Self = Self::ONE;
+
+    const TWO: Self = Self::TWO;
+
+    const MAX: Self = Self::MAX;
+}
+
 #[cfg(test)]
 mod tests {
+    use std::panic::catch_unwind;
+
+    use rand::Rng;
+
     use super::*;
 
+    fn u64_with_odd_bits_set() -> u64 {
+        let mut v = 0u64;
+
+        for i in (1..=63).step_by(2) {
+            v |= 1u64 << i;
+        }
+
+        v
+    }
+
+    fn u64_with_even_bits_set() -> u64 {
+        let mut v = 0u64;
+
+        // bit index are from 0 to 63
+        for i in (0..=62).step_by(2) {
+            v |= 1u64 << i;
+        }
+
+        v
+    }
+
+    #[test]
+    fn test_u64_even_odd_bits() {
+        let all_even_bits_set = u64_with_even_bits_set();
+        let all_odd_bits_set = u64_with_odd_bits_set();
+
+        assert_ne!(all_odd_bits_set, all_even_bits_set);
+
+        assert_eq!(all_even_bits_set.rotate_right(1), all_odd_bits_set);
+        assert_eq!(all_even_bits_set, all_odd_bits_set.rotate_left(1));
+    }
+
+    #[test]
+    fn test_bitand() {
+        let all_even_bits_set = U256([u64_with_even_bits_set(); 4]);
+        let all_odd_bits_set = U256([u64_with_odd_bits_set(); 4]);
+
+        assert_ne!(all_odd_bits_set, all_even_bits_set);
+        assert_eq!(all_odd_bits_set & all_odd_bits_set, all_odd_bits_set);
+        assert_eq!(all_even_bits_set & all_even_bits_set, all_even_bits_set);
+        assert_eq!(all_even_bits_set & all_odd_bits_set, U256::ZERO);
+    }
+
+    #[test]
+    fn test_bitor() {
+        let all_even_bits_set = U256([u64_with_even_bits_set(); 4]);
+        let all_odd_bits_set = U256([u64_with_odd_bits_set(); 4]);
+
+        assert_ne!(all_odd_bits_set, all_even_bits_set);
+        assert_eq!(all_odd_bits_set | all_odd_bits_set, all_odd_bits_set);
+        assert_eq!(all_even_bits_set | all_even_bits_set, all_even_bits_set);
+        assert_eq!(all_even_bits_set | all_odd_bits_set, U256::MAX);
+    }
+
+    #[test]
+    fn test_bitxor() {
+        let all_even_bits_set = U256([u64_with_even_bits_set(); 4]);
+        let all_odd_bits_set = U256([u64_with_odd_bits_set(); 4]);
+
+        assert_ne!(all_odd_bits_set, all_even_bits_set);
+        assert_eq!(all_odd_bits_set ^ all_odd_bits_set, U256::ZERO);
+        assert_eq!(all_even_bits_set ^ all_even_bits_set, U256::ZERO);
+        assert_eq!(all_even_bits_set ^ all_odd_bits_set, U256::MAX);
+    }
+
+    #[test]
+    fn test_is_power_of_two() {
+        assert!(!U256::ZERO.is_power_of_two());
+        assert!(!U256::MAX.is_power_of_two());
+        assert!(!U256::from(8329842348123u64).is_power_of_two());
+
+        assert!(U256::ONE.is_power_of_two());
+        assert!(U256::from(2u64).is_power_of_two());
+        assert!(U256::from(1u64 << 32).is_power_of_two());
+        assert!(U256::from(1u64 << 47).is_power_of_two());
+        assert!(U256::from(1u128 << 84).is_power_of_two());
```

Comment:
can we do them all in a few loops maybe ?

Reply:
I think yes